--- memcached.pm.orig   2009-09-11 18:19:22.000000000 +0700
+++ memcached.pm        2009-11-07 23:24:16.000000000 +0600
@@ -35,12 +35,12 @@

 sub store {
     my $self = shift;
-    my ($sid, $datastr) = @_;
+    my ($sid, $datastr,$etime) = @_;
     croak "store(): usage error" unless $sid && $datastr;

 #warn "store(): sid=$sid, $datastr\n";
     my $memcached = $self->{Memcached};
-    $memcached->set($sid, $datastr);
+    $memcached->set($sid, $datastr,$etime);

     return 1;
 }

